video/out/vo_drm.c

```diff
@@ -0,0 +1,523 @@
+                               struct modeset_dev **out)
+{
+    struct modeset_dev *dev = NULL;
+    drmModeConnector *conn = NULL;
+    drmModeRes *res = NULL;
+    int ret = 0;
+
+    *out = NULL;
+
+    res = drmModeGetResources(fd);
+    if (!res) {
+        char *errstr = mp_strerror(errno);
+        MP_ERR(vo, "Cannot retrieve DRM resources: %s\n", errstr);
+        ret = -errno;
+        goto end;
+    }
+
+    if (conn_id == -1) {
+        // get the first connected connector
+        for (int i = 0; i < res->count_connectors; i ++) {
+            conn = drmModeGetConnector(fd, res->connectors[i]);
+            if (is_connector_valid(vo, i, conn, true)) {
+                conn_id = i;
+                break;
+            }
+            if (conn) {
+                drmModeFreeConnector(conn);
+                conn = NULL;
+            }
+        }
+        if (conn_id == -1) {
+            MP_ERR(vo, "No connected connectors found\n");
+            ret = -ENODEV;
+            goto end;
+        }
+    }
+
+    if (conn_id < 0 || conn_id >= res->count_connectors) {
+        MP_ERR(vo, "Bad connector ID. Max valid connector ID = %u\n",
+               res->count_connectors);
+        ret = -ENODEV;
+        goto end;
+    }
+
+    conn = drmModeGetConnector(fd, res->connectors[conn_id]);
+    if (!is_connector_valid(vo, conn_id, conn, true)) {
+        ret = -ENODEV;
+        goto end;
+    }
+
+    dev = talloc_size(vo->priv, sizeof(*dev));
+    dev->conn = conn->connector_id;
+    dev->front_buf = 0;
+    memcpy(&dev->mode, &conn->modes[0], sizeof(dev->mode));
+    dev->bufs[0].width = conn->modes[0].hdisplay;
+    dev->bufs[0].height = conn->modes[0].vdisplay;
+    dev->bufs[1].width = conn->modes[0].hdisplay;
+    dev->bufs[1].height = conn->modes[0].vdisplay;
+
+    MP_INFO(vo, "Connector using mode %ux%u\n",
+            dev->bufs[0].width, dev->bufs[0].height);
+
+    ret = modeset_find_crtc(vo, fd, res, conn, dev);
+    if (ret) {
+        MP_ERR(vo, "Connector %d has no valid CRTC\n", conn_id);
+        goto end;
+    }
+
+    unsigned int i, j;
+    for (i = 0; i < BUF_COUNT; i ++) {
+        ret = modeset_create_fb(vo, fd, &dev->bufs[i]);
+        if (ret) {
+            for (j = 0; j < i; j ++) {
+                modeset_destroy_fb(fd, &dev->bufs[j]);
+            }
+            MP_ERR(vo, "Cannot create framebuffer for connector %d\n",
+                   conn_id);
+            return ret;
+        }
+    }
+
+end:
+    if (conn) {
+        drmModeFreeConnector(conn);
+        conn = NULL;
+    }
+    if (res) {
+        drmModeFreeResources(res);
+        res = NULL;
+    }
+    if (ret == 0) {
+        *out = dev;
+    } else {
+        talloc_free(dev);
+    }
+    return ret;
+}
+
+
+
+static int reconfig(struct vo *vo, struct mp_image_params *params, int flags)
+{
+    struct priv *p = vo->priv;
+    float device_ar = p->device_w / (float)p->device_h;
+
+    vo_get_src_dst_rects(vo, &p->src, &p->dst, &p->osd);
```

Comment:
The calculations below are unnecessarily complicated, because this isn't used correctly. ``vo->dwidth`` and ``vo->dheight`` should be set to the screen size before calling this function. (This is basically a leftover from ancient times - it would just be better if these were arguments to the ``vo_get_src_dst_rects()`` function.) Then p->dst is guaranteed to be within the screen size.